offchain/authority-claimer/src/listener.rs

```diff
@@ -2,61 +2,189 @@
 // SPDX-License-Identifier: Apache-2.0 (see LICENSE)
 
 use async_trait::async_trait;
-use rollups_events::{BrokerConfig, DAppMetadata, RollupsClaim};
-use snafu::Snafu;
+use rollups_events::{
+    Broker, BrokerConfig, BrokerError, DAppMetadata, RollupsClaimsStream,
+    INITIAL_ID,
+};
+use snafu::ResultExt;
 use std::fmt::Debug;
+use tracing::trace;
 
-use crate::metrics::AuthorityClaimerMetrics;
+use crate::claimer::Claimer;
 
-/// The `BrokerListener` listens for new claims from the broker.
-///
-/// The `listen` function should preferably yield to other processes while
-/// waiting for new messages (instead of busy-waiting).
+/// The `BrokerListener` starts a perpetual loop that listens for new claims from
+/// the broker and sends them to be processed by the injected `Claimer`.
 #[async_trait]
-pub trait BrokerListener: Debug {
-    type Error: snafu::Error;
+pub trait BrokerListener<C: Claimer> {
+    type Error: snafu::Error + 'static;
 
-    async fn listen(&self) -> Result<RollupsClaim, Self::Error>;
+    /// Starts the polling loop.
+    async fn start(self, claimer: C) -> Result<(), Self::Error>;
```

Comment:
We should revert to the original idea of the Claimer owning the Listener. I wonder why Renan made this change. Probably, the function signature needs to be:
```
async fn listen(&mut self) -> Reuslt<RollupsClaim, Self::Error>;
```

Reply:
Done at https://github.com/cartesi/rollups-node/commit/226c386d96f9c8445dea95a31c03455991bc9114